File: api/src/repositories/recipe/recipe-repository.ts
import SQL from 'sql-template-tag';
import { APIError404 } from '../../utils/error.js';
import { Repository } from '../repository.js';

const test = 'test';

/**
 * Recipe Repository class.
 *
 * @class RecipeRepository
 * @extends Repository
 */
export class RecipeRepository extends Repository {
  /**
   * Get recipe by id.
   *
   * @async
   * @param {number} recipeId
   * @throws {APIError404} - Recipe not found
   * @returns {IRecipeRecord}
   */
  async getRecipeById(recipeId: number): Promise<IRecipeRecord> {
    const sqlStatement = SQL`
      SELECT
        recipe_id,
        cookbook_id,
        name,
        url,
        description
      FROM recipe
      WHERE recipe_id = ${recipeId};
    `;

    const response = await this.connection.sql(sqlStatement);

    if (!response.length) {
      throw new APIError404('Recipe not found.', ['RecipeRepository->getRecipeById']);
    }

    return response[0];
  }

  /**
   * Get recipes by cookbook id.
   *
   * @async
   * @param {number} cookbookId
   * @returns {Promise<IRecipeRecord[]>}
   */
  async getRecipesByCookbookId(cookbookId: number): Promise<IRecipeRecord[]> {
    const sqlStatement = SQL`
      SELECT
        recipe_id,
        cookbook_id,
        name,
        url,
        description
      FROM recipe
      WHERE cookbook_id = ${cookbookId};
    `;

    return this.connection.sql(sqlStatement);
  }
}

File: backend/api/repository/recipeRepository.go
package repository

import (
	"database/sql"
	"recipevault/models"
)

type RecipeRepository struct {
	db *sql.DB
}

func NewRecipeRepository(db *sql.DB) *RecipeRepository {
	return &RecipeRepository{
		db: db,
	}
}

// Get Recipe by ID
func (r *RecipeRepository) GetRecipeByID(recipeID int) (models.Recipe, error) {
	query := `
    SELECT
      recipe_id,
      cookbook_id,
      name,
      url,
      description
    FROM recipe
    WHERE recipe_id = $1;`

	return GetRow[models.Recipe](r.db, query, recipeID)
}

// Get Recipes by Cookbook ID
func (r *RecipeRepository) GetRecipesByCookbookID(cookbookID int) ([]models.Recipe, error) {
	query := `
    SELECT
      r.recipe_id,
      r.cookbook_id,
      r.name,
      r.url,
      r.description
    FROM recipe r
    JOIN cookbook c
    ON r.cookbook_id = c.cookbook_id
    WHERE r.cookbook_id = $1;`

	return FindRows[models.Recipe](r.db, query, cookbookID)
}
